### ISSUE TITLE
[Linux] trackbar does not display the max value

### ISSUE THREAD

**Issue by ebachard** [Dovyski/cvui#18]
Hi Fernando,

First over all, congratulations for the 2.0 : the cmake is perfect, and the Linux build works without any glitch. Wow ! 

Second, your image buttons and all the improvement you did (everything in a .h ) desserve another wow too :-)

I only found one strange issue (see the attached screenshot) : the trackbar do not display the max value. I started to search, but I'm unsure, so I'll add more information, and let you know when things will be ok , means when I'll have a fix or at least a workaround (probably soemthing linked to sprintf_s replaced by sprintf not enough, but I'm unsure).

BTW, what about add those lines in the CMakeLists.txt file (located in $ROOT_DIR) ?

[diff_CMakelists.diff.txt](https://github.com/Dovyski/cvui/files/978968/diff_CMakelists.diff.txt)

Thanks again for sharing your code,

-- ericb

**Comment by ebachard** [comment]
![cvui_test_slider_no_max_value](https://cloud.githubusercontent.com/assets/23149049/25745316/a6fd1b82-319e-11e7-9ebc-2ceb541ad71d.png)

A screenshot of the issue on Linux

**Comment by ebachard** [comment]
A screenshot of the issue on Linux (look at the red ellipse)

[cvui_test_slider_no_max_value.pdf](https://github.com/Dovyski/cvui/files/978971/cvui_test_slider_no_max_value.pdf)

**Comment by pthom** [comment]
Hello,
I posted a PR that solved it under OS X
https://github.com/Dovyski/cvui/pull/15
Can you test it ?

**Comment by ebachard** [comment]
@pthom 
Hello. Thanks a lot for your patch and your support, greatly appreciated. To answer you, I applied all your changes manualy, and I confirm the current issue is fixed. => OK 

However, I still have the same problem with the other examples (I'll attach a screenshot of Trackbar example to illustrate). On your side, do you see eveything fixed, or is there yet a bit of little formating to fix for you ?

**Comment by ebachard** [comment]
![trackbar_some_max_values_not_displayed](https://cloud.githubusercontent.com/assets/23149049/25746277/0c4af64a-31a3-11e7-9de6-f7aed9848435.png)

**Comment by ebachard** [comment]
Looks like I found another little one: 
[trackbar_changes.diff.txt](https://github.com/Dovyski/cvui/files/979072/trackbar_changes.diff.txt)

**Comment by pthom** [comment]
Hum, I think I corrected this example also in the PR. If you did it manually, you might have skipped some needed modifications
Let me explain you how to get the changes without having to apply them manually :-)

```
# Revert all changes
git reset --hard 
# Add a new remote that point to the repo where the PR (Pull Request) resides
git remote add pthom https://github.com/pthom/cvui.git
# Apply the changes from the branch that corrects the formatting issue
git pull pthom/long_double_format
```

For you own use, the important message is : use "%lF" instead of "%lf" for the trackbar formatting options.

**Comment by ebachard** [comment]
@phtom

Yes, you're right, I just applied partialy the changes manualy (I'm used to patch language since a long time ;) ), but the problem is IMHO elsewhere : applying all your changes gives me several "nan" (not a number) with some trackbars (but not all, because there are several templates in the cvui.h).

That's the reason why I stepped backward and restarted from the begining, applying progressively the changes you propose. Currently - for me- the issue is not solved and I didn't figure out where exactly the problem stands. The wrong formatting in the string is certainly a good track, but I wonder whether this is the only reason. FYI, in meantime, I tested other changes in the whole API, modifying the templates, using float only, and I found other problems.  I don't even know whether this is a _Linux_only_ problem, or a design issue or something else (?). I'm not in a hurry and my time is limited, so I'll continue to search, just in case I find a track. HTH you to understand what I'm doing.

However, thanks a lot for the advices and the "git reset --hard " I didn't know. Instead, I'm used to type other commands like " git diff (file or path) | patch-R -pn and other things close. It seems to work too, but maybe what I do is wrong. Last but not least, I'll keep the second one (git pull pthom/long_double_format) preciously in my logs, because it's very usefull with git and can be applied for other projects.

-- 
ericb

**Comment by pthom** [comment]
Ok, I was able to reproduce the problem under linux. Working on it ...

**Comment by pthom** [comment]
Ok, I found the problem and solved it : 
Look at https://github.com/pthom/cvui/commit/84f9b11d8ed65f51f6553af8f1112d492c3ac00f
You can get the patch using `git pull pthom/long_double_format`

**Comment by ebachard** [comment]
@pthom : I got the same fix :-),  and I confirm the issue is fixed.

Thanks a lot !  I got it all working now :  Wow !

For the record, I was not able to ckeckout the change you did (I'm not handy using git, and got some rights issue), but I'll attach the full diff containing all te changes I did to make it work. It should be very similar to your changes (excepted the - currently useless - add_compile_options(-D__Linux) I added in the CMakefile.txt).

If your changes are accepted, please tell me, and I'll checkout a clean repo, to confirm the issue is definitely fixed.


[cvui2.0.0_full_diff.diff.txt](https://github.com/Dovyski/cvui/files/981647/cvui2.0.0_full_diff.diff.txt)

**Comment by Dovyski** [comment]
Thanks a lot for your hard work on this issue, guys! I will do a final check of the fix on Windows and merge the pull request (within the next three days).

**Comment by Dovyski** [comment]
I've just merged the pull request regarding this issue. It breaks on Windows (Visual Studio 2015) because of the `static_cast` in the `trackbar()` function definition:

`bool trackbar(int theWidth, T *theValue, T theMin, T theMax, int theSegments = 1, const char *theLabelFormat = "%.1Lf", unsigned int theOptions = 0, T theDiscreteStep = static_cast<T>(1.));`

If `static_cast` is not used, e.g. 

`bool trackbar(int theWidth, T *theValue, T theMin, T theMax, int theSegments = 1, const char *theLabelFormat = "%.1Lf", unsigned int theOptions = 0, T theDiscreteStep = 1);`

things compile and work fine on Windows. The use of `static_cast` looks necessary to me, but is it really required for a compilation on OSX and Linux? It looks like MSVC started the witch hunting season :smile:

**Comment by pthom** [comment]
Hi Fernando, 

My opinion is that the static_cast ought to be necessary. However, if you remove the static_cast in the declarations only it will still compile and run correctly under linux and OSX. I just made the test and it is ok.

It is probably an issue with visual studio 2015 : static_cast<double>(1.) should be a constexpr, be it seems that visual studio does not see it this way.

So, I guess you can remove it.

**Comment by ebachard** [comment]
Looks like this issue is very close to be closed !

strange issue on Windows, but I'm not surprised  ;-) Of course remove -imho correct - static_cast is no problem for me, and I confirm a simple " (long double)theValue " works on Linux too. Not tested on Mac OS X though.
 +... I forgot to mention one format issue in row-coluwn/main.cpp 
... and last, other possibility to assume portability could be to use a little -D__Linux in the CMakeLists.txt file. (see attached files)

[CMakeLists.txt.diff.txt](https://github.com/Dovyski/cvui/files/984575/CMakeLists.txt.diff.txt)
[row_column_cvui#18.diff.txt](https://github.com/Dovyski/cvui/files/984576/row_column_cvui.18.diff.txt)

@pthom : how does it work on the pixels side with hdpi on Mac OS X ? My current resolution is 1680 X 1050 (Linux on macbook pro 6,2 , 2011 model + " antiglare" screen), and I never tested more using cvui

**Comment by Dovyski** [comment]
I think everything is sorted out, guys. Commit 0a3d61bbee01813d8caba6aa6a99808d0bb4fa70 merges all fixes into master. I've created the release [v2.0.0-RC1](https://github.com/Dovyski/cvui/releases/tag/v2.0.0-RC1) to check if everything is ok. Could you please check if this version works for you?

**Comment by ebachard** [comment]
Hi Fernando,

I tested everything, and I see no problem => see the full report I added below. 
Thanks agains for sharing your code !

[cvui_issue#18_report.txt](https://github.com/Dovyski/cvui/files/1003681/cvui_issue.18_report.txt)

**Comment by pthom** [comment]
Hi,

Everything ok on my side after my tests under OSX and linux.

I only saw a minor warning while compiling under OSX :
```
cvui-2.0.0-RC1/example/src/trackbar-sparkline/main.cpp:46:44: warning: implicit conversion from
      'double' to 'int' changes value from 0.25 to 0 [-Wliteral-conversion]
                        if (cvui::trackbar(150, &value, 0., 5., 0.25)) {
                            ~~~~                                ^~~~
```

This warning can be silenced by removing the last param (0.25).

For the rest, perfect :-)

**Comment by Dovyski** [comment]
Thank you guys for such valuable help (and awesome reports)! :smile:
I will fix the OSX warning and ship version 2.0.

**Comment by Dovyski** [comment]
Again thanks for all the help, guys! This issue has been fixed and [v2.0.0](https://github.com/Dovyski/cvui/releases/tag/v2.0.0) is up.
